pkg/ingester/instance.go

```diff
@@ -108,6 +109,12 @@ type instance struct {
 	streamRateCalculator *StreamRateCalculator
 
 	writeFailures *writefailures.Manager
+
+	schemaconfig *config.SchemaConfig
+
+	// precomputed `Chunk` and `Head` formats parsed from schemaconfig
+	chunkfmt byte
```

Comment:
This should be detected when writing a chunk within the `stream`, not on the instance. The `instance` may be exist across schema boundaries. This leaves us with two choices:
1) set the format based on the min timestamp in a chunk when it's cut
2) set the format based on the max timestamp in a chunk when it's cut

I think (1) is the more conservative approach and thus more appropriate. We may write an older format to a newer schema bucket, but we won't write a newer format to an older bucket. To be exact, the chunk is written once then added to the index buckets for different days. I'd rather us have some v13-schema indices reference v12 compatible chunks than vice versa.

Reply:
Nice catch. thanks @owen-d for the explanation. 

In fact [my initial commit used "per-stream" chunk format selection](https://github.com/grafana/loki/pull/10213/commits/8b6787fcb8c7d75984be131ea31faeec184af5db#diff-5e8ea6dc696f5e7c429726e1d8c59705869c8a64e1793f399e6e6587b137e4bfL296-L302), but was using `model.Now()` instead of `min/max` timestamp. Makes more sense to use correct timestamp here.

Also agree with `min` timestamp.